File: src/0322-Coin-Change/0322.go
var res int64

func coinChange(coins []int, amount int) int {
    sort.Ints(coins)
    res = int64(amount) + 1

    dfs(coins, len(coins) - 1, int64(amount), 0)
    if res > int64(amount) {
        return -1
    }
    return int(res)
}

func dfs(coins []int, index int, cur, cnt int64) {
    coins_i := int64(coins[index])
    if cnt + (cur + coins_i - 1) / coins_i >= res {
        return 
    }

    if cur % coins_i == 0 {
        res = cnt + cur / coins_i
        return
    }

    if index == 0 {
        return
    }

    for i := cur / coins_i; i >= 0; i-- {
        dfs(coins, index - 1, cur - coins_i * i, cnt + i)
    }
}

File: src/0322-Coin-Change/0322.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Arrays.sort(coins);
        res = amount + 1;
        dfs(coins, coins.length - 1, amount, 0);
        return (int)res > amount ? -1 : (int)res;
    }

    private long res;

    private void dfs(int[] coins, int index, long cur, long cnt) {
        long coins_i = coins[index];
        if (cnt + (cur + coins_i - 1) / coins_i >= res) return;

        if (cur % coins_i == 0) {
            res = cnt + cur / coins_i;
            return ;
        } 

        if (index == 0) return ;

        for (long i = cur / coins_i; i >= 0; i--) 
            dfs(coins, index - 1, cur - coins_i * i, cnt + i);
    }
}